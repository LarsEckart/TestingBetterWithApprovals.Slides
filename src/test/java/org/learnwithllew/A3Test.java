package org.learnwithllew;

import org.apache.commons.lang3.mutable.MutableInt;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.List;

public class A3Test {

    @Test
    void testConversations() {
        var conversations = List.of(
            conversation("hi"),
            conversation("hi", "hi"),
            conversation("hi", "hi", "hi"),
            conversation("hi", "pay bill"),
            conversation("hi", "pay bill", "Yes, I'm a customer"),
            conversation("hi", "pay bill", "no"),
            conversation("hi", "talk to an operator", "Yes, I'm a customer"),
            conversation("pay bill"),
            conversation("pay bill", "I like swimming"),
            conversation("pay bill", "I like coffee", "I like tea", "no"),
            conversation("pay bill", "Yes, I'm a customer")
                .and("pay bill"),
            conversation("pay bill", "No, I'm not")
                .and("pay bill"),

            conversation("talk to an operator"),
            conversation("oh hi there, how are you doing"),
            conversation("walk my dog", "Yes, I'm a customer")
//            conversation("pay bill")
//                .withCustomer(CustomerType.EXISTING)
//                .duringWorkHours()
//                .and("pay bill")
//                .duringOffHours()
        );

        MutableInt counter = new MutableInt(0);
        Approvals.verifyAll("Chatbot conversations", conversations, c -> haveConversation(c, counter));
    }

    private String haveConversation(Conversations conversations, MutableInt counter) {
        counter.increment();
        BotOutput output = new BotOutput();
        Bot bot = new Bot(output);

        var storyBoard = "# Scenario %s: %s\n".formatted(counter, conversations.printMessages());
        for (int i = 0; i < conversations.conversations.size(); i++) {
            var conversation = conversations.conversations.get(i);
            var messages = conversation.messages;
            storyBoard += String.format("%s***** Conversation %s *****\n", i == 0 ? "" : "\n", i + 1);
            storyBoard += StoryBoard.create(bot, output, messages);
        }
        return storyBoard;
    }

    private Conversations conversation(String... messages) {
        return new Conversations(messages);
    }
}
