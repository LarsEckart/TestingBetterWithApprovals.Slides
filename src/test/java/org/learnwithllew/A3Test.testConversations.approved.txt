Chatbot conversations


# Scenario 1: [hi]
***** Conversation 1 *****
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?

# Scenario 2: [hi, hi]
***** Conversation 1 *****
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: hi
[     Bot]: Hmmm, tell me a little more so I can help you.

What would you like to do today?

# Scenario 3: [hi, hi, hi]
***** Conversation 1 *****
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: hi
[     Bot]: Hmmm, tell me a little more so I can help you.

What would you like to do today?
[Customer]: hi
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not

# Scenario 4: [hi, pay bill]
***** Conversation 1 *****
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: pay bill
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not

# Scenario 5: [hi, pay bill, Yes, I'm a customer]
***** Conversation 1 *****
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: pay bill
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: Yes, I'm a customer
[     Bot]: transfers to 'self_service'

# Scenario 6: [hi, pay bill, no]
***** Conversation 1 *****
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: pay bill
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: no
[     Bot]: transfers to 'operator'

# Scenario 7: [hi, talk to an operator, Yes, I'm a customer]
***** Conversation 1 *****
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: talk to an operator
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: Yes, I'm a customer
[     Bot]: transfers to 'operator'

